Src/LiquidProjections.NEventStore/Subscription.cs

```diff
@@ -44,19 +45,109 @@ public void Start()
                         TaskCreationOptions.DenyChildAttach | TaskCreationOptions.LongRunning,
                         TaskScheduler.Default)
                     .Unwrap();
             }
         }
 
+        private async Task RunAsync(SubscriptionInfo info)
+        {
+            const int offsetToDetectAheadSubscriber = 1;
+            long actualRequestedLastCheckpoint = lastProcessedCheckpoint;
+            lastProcessedCheckpoint = lastProcessedCheckpoint > 0 ? lastProcessedCheckpoint - offsetToDetectAheadSubscriber : 0;
+            bool firstRequestAfterSubscribing = true;
+
+            while (!cancellationTokenSource.IsCancellationRequested)
+            {
+#if DEBUG
+                LogProvider.GetCurrentClassLogger().Debug(() =>
+                    $"Subscription {Id} is requesting a page after checkpoint {lastProcessedCheckpoint}.");
+#endif
+                Page page = await TryGetNextPage();
+                if (page != null)
+                {
+#if DEBUG
+                    LogProvider.GetCurrentClassLogger().Debug(() =>
+                        $"Subscription {Id} has got a page of size {page.Transactions.Count} " +
+                        $"from checkpoint {page.Transactions.First().Checkpoint} " +
+                        $"to checkpoint {page.Transactions.Last().Checkpoint}.");
+#endif
+
+                    if (firstRequestAfterSubscribing && !page.Transactions.Any())
+                    {
+                        await subscriber.NoSuchCheckpoint(info);
+                    }
+                    else
+                    {
+                        var transactions = page.Transactions
+                            .Where(t => t.Checkpoint > actualRequestedLastCheckpoint)
+                            .ToReadOnly();
+
+                        if (transactions.Count > 0)
+                        {
+                            await subscriber.HandleTransactions(transactions, info).ConfigureAwait(false);
+                        }
+                    }
+
+                    firstRequestAfterSubscribing = false;
+
+#if DEBUG
+                    LogProvider.GetCurrentClassLogger().Debug(() =>
+                        $"Subscription {Id} has processed a page of size {page.Transactions.Count} " +
+                        $"from checkpoint {page.Transactions.First().Checkpoint} " +
+                        $"to checkpoint {page.Transactions.Last().Checkpoint}.");
+#endif
+
+                    lastProcessedCheckpoint = page.LastCheckpoint;
+                }
+            }
+        }
+
+        private async Task<Page> TryGetNextPage()
+        {
+            Page page = null;
+
+            try
+            {
+                page = await eventStoreAdapter.GetNextPage(lastProcessedCheckpoint, Id)
+                    .WithWaitCancellation(cancellationTokenSource.Token)
+                    .ConfigureAwait(false);
+            }
+            catch
+            {
+                // Just continue the next iteration
```

Comment:
Yes, but there can theoretically be other exceptions when getting the next page, coming not from the NEventStore. As an alternative, we could wrap the NEventStore exception to our own exception and only catch that one here.